diff --git a/include/mpark/variant.hpp b/include/mpark/variant.hpp
index 2fb2ac549..76f58530a 100644
--- a/include/mpark/variant.hpp
+++ b/include/mpark/variant.hpp
@@ -245,7 +245,7 @@ namespace mpark {
   };
 
   [[noreturn]] inline void throw_bad_variant_access() {
-#ifdef MPARK_EXCEPTIONS
+#if defined(MPARK_EXCEPTIONS) && !defined(__CUDACC__)
     throw bad_variant_access{};
 #else
     std::terminate();
@@ -1021,9 +1021,15 @@ namespace mpark {
     friend struct access::recursive_union;                                 \
   }
 
+#ifdef __CUDACC__
+#define V_GPU_FUNCTION __host__ __device__
+#else
+#define V_GPU_FUNCTION
+#endif
+
     MPARK_VARIANT_RECURSIVE_UNION(Trait::TriviallyAvailable,
                                   ~recursive_union() = default;);
-    MPARK_VARIANT_RECURSIVE_UNION(Trait::Available,
+    MPARK_VARIANT_RECURSIVE_UNION(Trait::Available,V_GPU_FUNCTION
                                   ~recursive_union() {});
     MPARK_VARIANT_RECURSIVE_UNION(Trait::Unavailable,
                                   ~recursive_union() = delete;);
@@ -1128,14 +1134,14 @@ namespace mpark {
     MPARK_VARIANT_DESTRUCTOR(
         Trait::TriviallyAvailable,
         ~destructor() = default;,
-        inline void destroy() noexcept {
+        inline constexpr void destroy() noexcept {
           this->index_ = static_cast<index_t<Ts...>>(-1);
         });
 
     MPARK_VARIANT_DESTRUCTOR(
-        Trait::Available,
+        Trait::Available,V_GPU_FUNCTION
         ~destructor() { destroy(); },
-        inline void destroy() noexcept {
+        inline constexpr void destroy() noexcept {
           if (!this->valueless_by_exception()) {
             visitation::alt::visit_alt(dtor{}, *this);
           }
@@ -1169,14 +1175,14 @@ namespace mpark {
 #endif
 
       template <std::size_t I, typename T, typename... Args>
-      inline static T &construct_alt(alt<I, T> &a, Args &&... args) {
+      inline constexpr static T &construct_alt(alt<I, T> &a, Args &&... args) {
         auto *result = ::new (static_cast<void *>(lib::addressof(a)))
             alt<I, T>(in_place_t{}, lib::forward<Args>(args)...);
         return result->value;
       }
 
       template <typename Rhs>
-      inline static void generic_construct(constructor &lhs, Rhs &&rhs) {
+      inline constexpr static void generic_construct(constructor &lhs, Rhs &&rhs) {
         lhs.destroy();
         if (!rhs.valueless_by_exception()) {
           visitation::alt::visit_alt_at(
@@ -1223,7 +1229,7 @@ namespace mpark {
 
     MPARK_VARIANT_MOVE_CONSTRUCTOR(
         Trait::Available,
-        move_constructor(move_constructor &&that) noexcept(
+        constexpr move_constructor(move_constructor &&that) noexcept(
             lib::all<std::is_nothrow_move_constructible<Ts>::value...>::value)
             : move_constructor(valueless_t{}) {
           this->generic_construct(*this, lib::move(that));
@@ -1261,7 +1267,7 @@ namespace mpark {
 
     MPARK_VARIANT_COPY_CONSTRUCTOR(
         Trait::Available,
-        copy_constructor(const copy_constructor &that)
+        constexpr copy_constructor(const copy_constructor &that)
             : copy_constructor(valueless_t{}) {
           this->generic_construct(*this, that);
         });
@@ -1281,7 +1287,7 @@ namespace mpark {
       using super::operator=;
 
       template <std::size_t I, typename... Args>
-      inline /* auto & */ auto emplace(Args &&... args)
+      inline constexpr /* auto & */ auto emplace(Args &&... args)
           -> decltype(this->construct_alt(access::base::get_alt<I>(*this),
                                           lib::forward<Args>(args)...)) {
         this->destroy();
@@ -1304,7 +1310,7 @@ namespace mpark {
 #endif
 
       template <std::size_t I, typename T, typename Arg>
-      inline void assign_alt(alt<I, T> &a, Arg &&arg) {
+      inline constexpr void assign_alt(alt<I, T> &a, Arg &&arg) {
         if (this->index() == I) {
 #ifdef _MSC_VER
 #pragma warning(push)
@@ -1332,7 +1338,7 @@ namespace mpark {
       }
 
       template <typename That>
-      inline void generic_assign(That &&that) {
+      inline constexpr void generic_assign(That &&that) {
         if (this->valueless_by_exception() && that.valueless_by_exception()) {
           // do nothing.
         } else if (that.valueless_by_exception()) {
@@ -1381,7 +1387,7 @@ namespace mpark {
 
     MPARK_VARIANT_MOVE_ASSIGNMENT(
         Trait::Available,
-        move_assignment &
+        constexpr move_assignment &
         operator=(move_assignment &&that) noexcept(
             lib::all<(std::is_nothrow_move_constructible<Ts>::value &&
                       std::is_nothrow_move_assignable<Ts>::value)...>::value) {
@@ -1394,6 +1400,7 @@ namespace mpark {
         move_assignment &operator=(move_assignment &&) = delete;);
 
 #undef MPARK_VARIANT_MOVE_ASSIGNMENT
+#undef V_GPU_FUNCTION
 
     template <typename Traits, Trait = Traits::copy_assignable_trait>
     class copy_assignment;
@@ -1421,7 +1428,7 @@ namespace mpark {
 
     MPARK_VARIANT_COPY_ASSIGNMENT(
         Trait::Available,
-        copy_assignment &operator=(const copy_assignment &that) {
+        constexpr copy_assignment &operator=(const copy_assignment &that) {
           this->generic_assign(that);
           return *this;
         });
@@ -1447,7 +1454,7 @@ namespace mpark {
       impl &operator=(impl &&) = default;
 
       template <std::size_t I, typename Arg>
-      inline void assign(Arg &&arg) {
+      inline constexpr void assign(Arg &&arg) {
         this->assign_alt(access::base::get_alt<I>(*this),
                          lib::forward<Arg>(arg));
       }
@@ -1499,7 +1506,7 @@ namespace mpark {
 #ifndef MPARK_GENERIC_LAMBDAS
       struct swapper {
         template <typename ThisAlt, typename ThatAlt>
-        inline void operator()(ThisAlt &this_alt, ThatAlt &that_alt) const {
+        inline constexpr void operator()(ThisAlt &this_alt, ThatAlt &that_alt) const {
           using std::swap;
           swap(this_alt.value, that_alt.value);
         }
@@ -1694,8 +1701,8 @@ namespace mpark {
 
     ~variant() = default;
 
-    variant &operator=(const variant &) = default;
-    variant &operator=(variant &&) = default;
+    constexpr variant &operator=(const variant &) = default;
+    constexpr variant &operator=(variant &&) = default;
 
     template <typename Arg,
               lib::enable_if_t<!std::is_same<lib::decay_t<Arg>, variant>::value,
@@ -1705,7 +1712,7 @@ namespace mpark {
               lib::enable_if_t<(std::is_assignable<T &, Arg>::value &&
                                 std::is_constructible<T, Arg>::value),
                                int> = 0>
-    inline variant &operator=(Arg &&arg) noexcept(
+    inline constexpr variant &operator=(Arg &&arg) noexcept(
         (std::is_nothrow_assignable<T &, Arg>::value &&
          std::is_nothrow_constructible<T, Arg>::value)) {
       impl_.template assign<I>(lib::forward<Arg>(arg));
@@ -1717,7 +1724,7 @@ namespace mpark {
         typename... Args,
         typename T = lib::type_pack_element_t<I, Ts...>,
         lib::enable_if_t<std::is_constructible<T, Args...>::value, int> = 0>
-    inline T &emplace(Args &&... args) {
+    inline constexpr T &emplace(Args &&... args) {
       return impl_.template emplace<I>(lib::forward<Args>(args)...);
     }
 
@@ -1730,7 +1737,7 @@ namespace mpark {
                                                std::initializer_list<Up> &,
                                                Args...>::value,
                          int> = 0>
-    inline T &emplace(std::initializer_list<Up> il, Args &&... args) {
+    inline constexpr T &emplace(std::initializer_list<Up> il, Args &&... args) {
       return impl_.template emplace<I>(il, lib::forward<Args>(args)...);
     }
 
@@ -1739,7 +1746,7 @@ namespace mpark {
         typename... Args,
         std::size_t I = detail::find_index_sfinae<T, Ts...>::value,
         lib::enable_if_t<std::is_constructible<T, Args...>::value, int> = 0>
-    inline T &emplace(Args &&... args) {
+    inline constexpr T &emplace(Args &&... args) {
       return impl_.template emplace<I>(lib::forward<Args>(args)...);
     }
 
@@ -1752,7 +1759,7 @@ namespace mpark {
                                                std::initializer_list<Up> &,
                                                Args...>::value,
                          int> = 0>
-    inline T &emplace(std::initializer_list<Up> il, Args &&... args) {
+    inline constexpr T &emplace(std::initializer_list<Up> il, Args &&... args) {
       return impl_.template emplace<I>(il, lib::forward<Args>(args)...);
     }
 
@@ -1772,7 +1779,7 @@ namespace mpark {
                             lib::dependent_type<lib::is_swappable<Ts>,
                                                 Dummy>::value)...>::value,
                   int> = 0>
-    inline void swap(variant &that) noexcept(
+    inline constexpr void swap(variant &that) noexcept(
         lib::all<(std::is_nothrow_move_constructible<Ts>::value &&
                   lib::is_nothrow_swappable<Ts>::value)...>::value) {
       impl_.swap(that.impl_);
@@ -2093,7 +2100,7 @@ namespace mpark {
 #endif
 
   template <typename... Ts>
-  inline auto swap(variant<Ts...> &lhs,
+  inline constexpr auto swap(variant<Ts...> &lhs,
                    variant<Ts...> &rhs) noexcept(noexcept(lhs.swap(rhs)))
       -> decltype(lhs.swap(rhs)) {
     lhs.swap(rhs);
